README.md
# Voctomix [![Build Status](https://travis-ci.org/voc/voctomix.svg?branch=master)](https://travis-ci.org/voc/voctomix)

**Hint**: Also see upcoming version 2 called *VOC2MIX* which can be found at [branch `voctomix2`](https://github.com/voc/voctomix/tree/voctomix2).

---

The [C3VOC](https://c3voc.de/) creates lecture recordings from German hacker/tech conferences. In the past we have used [dvswitch](http://dvswitch.alioth.debian.org/wiki/) very successfully but it has some serious limitations. Therefore we started looking for a replacement in 2014. We tested [snowmix](http://sourceforge.net/projects/snowmix/) and [gst-switch](https://github.com/timvideos/gst-switch) and while both did some things we wanted right, we realised that no existing tool would be able to fulfil all our wishes. Furthermore both are a nightmare to extend. So we decided to build our own implementation of a Live-Video-Mixer.

## Subprojects
The Voctomix project consists of three parts:
 - [Voctocore](./voctocore/), the videomixer core-process that does the actual video- and audio crunching
 - [Voctogui](./voctogui/), a GUI implementation in GTK controlling the core's functionality and giving visual feedback of the mixed video
 - Voctomix [Example Scripts](./example-scripts/), a collection of tools and examples for talking to the core-process, feeding and receiving video-streams and controlling operations from scripts or command-line.

## Installation
Voctomix requires a fairly recent Version of GStreamer (at least 1.5, though we recommend 1.6 and later). This is natively present on [Debian Stretch](https://packages.debian.org/stretch/libgstreamer1.0-0) or [Sid](https://packages.debian.org/sid/libgstreamer1.0-0) and [Ubuntu Wily](http://packages.ubuntu.com/wily/libgstreamer1.0-0). On these systems it should run out of the box. We recommend using one of them. A [Docker](http://www.docker.com) image that uses Ubuntu Wily as a base is bundled with Voctomix. Please refer to the seperate [readme](./README_DOCKER.md) how to use the Docker image.

Install the required dependencies:
````
# Requirements
apt-get install gstreamer1.0-plugins-bad gstreamer1.0-plugins-base gstreamer1.0-plugins-good gstreamer1.0-plugins-ugly gstreamer1.0-tools libgstreamer1.0-0 python3 python3-gi gir1.2-gstreamer-1.0 gir1.2-gst-plugins-base-1.0

# Optional for the Example-Scripts
apt-get install python3-pyinotify gstreamer1.0-libav rlwrap fbset ffmpeg netcat gstreamer1.0-vaapi
````


For the GUI you'll -- additionally to a gnome-desktop -- need to install the following dependencies:
````
apt-get install gstreamer1.0-plugins-bad gstreamer1.0-plugins-base gstreamer1.0-plugins-good gstreamer1.0-plugins-ugly gstreamer1.0-alsa gstreamer1.0-tools libgstreamer1.0-0 python3 python3-gi python3-gi-cairo gir1.2-gstreamer-1.0 gir1.2-gst-plugins-base-1.0 gir1.2-gtk-3.0
````

Now you should be able to clone the git-repository and run Voctomix or the GUI like this:
````
git clone https://github.com/voc/voctomix.git
cd voctomix
./voctocore/voctocore.py -vv
./voctogui/voctogui.py -vv
````

## Quickstart using Docker

### docker-compose

Install docker-compose (prerequisite)
```
apt install docker-compose      # debian
dnf install docker-compose      # fedora
```

Run the core, gui and two example source streams
```
xhost +local:$(id -un)
mkdir /tmp/vocto/ && touch /tmp/vocto/configgui.ini
GID=$(id -g) UID=$(id -u) docker-compose up
```

Run only the core
```
docker-compose up voctocore
```

Clean up stale containers
```
docker-compose rm
```

### Manually

Run the core and two example source streams
```
docker run -it --rm --name=voctocore c3voc/voctomix core
docker run -it --rm --name=cam1 --link=voctocore:corehost c3voc/voctomix gstreamer/source-videotestsrc-as-cam1.sh
docker run -it --rm --name=bg --link=voctocore:corehost c3voc/voctomix gstreamer/source-videotestsrc-as-background-loop.sh
```

Run the GUI 
```
xhost +local:$(id -un)
mkdir /tmp/vocto/ && touch /tmp/vocto/configgui.ini
docker run -it --rm --name=gui --env=gid=$(id -g) --env=uid=$(id -u) --env=DISPLAY=:0 --link=voctocore:corehost \
  -v /tmp/vocto/configgui.ini:/opt/voctomix/voctogui/config.ini -v /tmp/.X11-unix:/tmp/.X11-unix -v /tmp/.docker.xauth:/tmp/.docker.xauth c3voc/voctomix gui
```

show more commands availabe
```
docker run -it --rm c3voc/voctomix help
docker run -it --rm c3voc/voctomix examples
```

## A word on CPU usage
Voctomix requires a fair amount of CPU time to run in the default configuration of 1920×1080 at 25fps. Our production systems have these CPUs: `Intel Core i7-3770 CPU 4x 3.40GHz` but we're also experimenting with newer ones like these: `Intel Core i7-6700K, 4x 4.00GHz`.
For testing and development you may want to use a `config.ini` that reduces the resolution and also turns off the JPEG preview encoders, which take a huge amount of the required CPU power and are not required, as long as the GUI and the Core run on the same machine (or have a 10GE Link between them, FWIW). Don't forget to modify your source scripts to provide the correct resolution.

Such a config.ini might look like this:
````
[mix]
videocaps=video/x-raw,format=UYVY,width=320,height=180,framerate=25/1,pixel-aspect-ratio=1/1

[previews]
enabled=false
videocaps=video/x-raw,width=320,height=180,framerate=25/1
````

## A word on running in a VM
While the Core runs fine inside any VM, the GUI uses OpenGL to display the video streams. Don't forget to enable 3D acceleration in your VM to support this.


## Contact
To get in touch with us we'd ask to join `#voctomix` on the hackint IRC network, mail to `voc AT c3voc DOT de` or meet us on one of the [many conferences](https://c3voc.de/eventkalender) we're at.
You may also want to watch MaZderMinds [talk](https://media.ccc.de/v/froscon2016-1696-voctomix) about Voctomix and [a presentation](https://media.ccc.de/v/froscon2015-1520-conference_recording_und_streaming) some of us gave about C3VOCs video infrastructure.


debian/changelog
voc2mix-isdn (0.1) unstable; urgency=medium

  * Initial release.

 -- derchris <derchris@me.com>  Wed, 05 Aug 2020 14:14:42 +0200


vocto/config.py
#!/usr/bin/env python3
import logging
import re
import os

from gi.repository import Gst
from configparser import SafeConfigParser
from lib.args import Args
from vocto.transitions import Composites, Transitions
from vocto.audio_streams import AudioStreams
from vocto import kind_has_audio, kind_has_video

testPatternCount = 0

GST_TYPE_VIDEO_TEST_SRC_PATTERN = [
    "smpte",
    "ball",
    "red",
    "green",
    "blue",
    "black",
    "white",
    "checkers-1",
    "checkers-2",
    "checkers-4",
    "checkers-8",
    "circular",
    "blink",
    "smpte75",
    "zone-plate",
    "gamut",
    "chroma-zone-plate",
    "solid-color",
    "smpte100",
    "bar",
    "snow",
    "pinwheel",
    "spokes",
    "gradient",
    "colors"
]

GST_TYPE_AUDIO_TEST_SRC_WAVE = [
    "sine",
    "square",
    "saw",
    "triangle",
    "silence",
    "white-noise",
    "pink-noise",
    "sine-table",
    "ticks",
    "gaussian-noise",
    "red-noise",
    "blue-noise",
    "violet-noise",
]


class VocConfigParser(SafeConfigParser):

    log = logging.getLogger('VocConfigParser')
    audio_streams = None

    def getList(self, section, option, fallback=None):
        if self.has_option(section, option):
            option = self.get(section, option).strip()
            if len(option) == 0:
                return []

            unfiltered = [x.strip() for x in option.split(',')]
            return list(filter(None, unfiltered))
        else:
            return fallback

    def getSources(self):
        return self.getList('mix', 'sources')

    def getLiveSources(self):
        return ["mix"] + self.getList('mix', 'livesources', [])

    def getBackgroundSources(self):
        if self.has_option('mix', 'backgrounds'):
            return self.getList('mix', 'backgrounds')
        elif self.has_section('source.background'):
            return ["background"]
        else:
            return []

    def getBackgroundSource(self,composite):
        if not self.getBackgroundSources():
            return None
        for source in self.getBackgroundSources():
            if composite in self.getList('source.{}'.format(source), 'composites', fallback=[]):
                return source
        return self.getBackgroundSources()[0]

    def getSourceKind(self, source):
        return self.get('source.{}'.format(source), 'kind', fallback='test')

    def getNoSignal(self):
        nosignal = self.get('mix', 'nosignal', fallback='smpte100').lower()
        if nosignal in ['none','false','no']:
            return None
        elif nosignal in GST_TYPE_VIDEO_TEST_SRC_PATTERN:
            return nosignal
        else:
            self.log.error("Configuration value mix/nosignal has unknown pattern '{}'".format(nosignal))

    def getDeckLinkDeviceNumber(self, source):
        return self.getint('source.{}'.format(source), 'devicenumber', fallback=0)

    def getDeckLinkAudioConnection(self, source):
        return self.get('source.{}'.format(source), 'audio_connection', fallback='auto')

    def getDeckLinkVideoConnection(self, source):
        return self.get('source.{}'.format(source), 'video_connection', fallback='auto')

    def getDeckLinkVideoMode(self, source):
        return self.get('source.{}'.format(source), 'video_mode', fallback='auto')

    def getDeckLinkVideoFormat(self, source):
        return self.get('source.{}'.format(source), 'video_format', fallback='auto')

    def getPulseAudioDevice(self, source):
        return self.get('source.{}'.format(source), 'device', fallback='auto')

    def getAlsaAudioDevice(self, source):
        return self.get('source.{}'.format(source), 'device', fallback='hw:0')

    def getV4l2Device(self, source):
        return self.get('source.{}'.format(source), 'device', fallback='/dev/video0')

    def getV4l2Type(self, source):
        return self.get('source.{}'.format(source), 'type', fallback='video/x-raw')

    def getV4l2Width(self, source):
        return self.get('source.{}'.format(source), 'width', fallback=1920)

    def getV4l2Height(self, source):
        return self.get('source.{}'.format(source), 'height', fallback=1080)

    def getV4l2Format(self, source):
        return self.get('source.{}'.format(source), 'format', fallback='YUY2')

    def getV4l2Framerate(self, source):
        return self.get('source.{}'.format(source), 'framerate', fallback='25/1')

    def getRPICamDevice(self, source):
        return self.get('source.{}'.format(source), 'device', fallback='/dev/video0')

    def getRPICamType(self, source):
        return self.get('source.{}'.format(source), 'type', fallback='video/x-raw')

    def getRPICamWidth(self, source):
        return self.get('source.{}'.format(source), 'width', fallback=1920)

    def getRPICamHeight(self, source):
        return self.get('source.{}'.format(source), 'height', fallback=1080)

    def getRPICamCrop(self, source):
        return self.get('source.{}'.format(source), 'crop', fallback=None)

    def getRPICamFormat(self, source):
        return self.get('source.{}'.format(source), 'format', fallback='YUY2')

    def getRPICamFramerate(self, source):
        return self.get('source.{}'.format(source), 'framerate', fallback='25/1')

    def getRPICamAnnotation(self, source):
        return self.get('source.{}'.format(source), 'annotation', fallback=None)

    def getImageURI(self,source):
        if self.has_option('source.{}'.format(source), 'imguri'):
            return self.get('source.{}'.format(source), 'imguri')
        else:
            path = os.path.abspath(self.get('source.{}'.format(source), 'file'))
            if not os.path.isfile(path):
                self.log.error("image file '%s' could not be found" % path)
            return "file://{}".format(path)

    def getLocation(self,source):
        return self.get('source.{}'.format(source), 'location')

    def getLoop(self,source):
        return self.get('source.{}'.format(source), 'loop', fallback="true")

    def getTestPattern(self, source):
        if not self.has_section('source.{}'.format(source)):
            # default blinder source shall be smpte (if not defined otherwise)
            if source == "blinder":
                return "smpte"
            # default background source shall be black (if not defined otherwise)
            if source in self.getBackgroundSources():
                return "black"

        pattern = self.get('source.{}'.format(source), 'pattern', fallback=None)
        if not pattern:
            global testPatternCount
            testPatternCount += 1
            pattern = GST_TYPE_VIDEO_TEST_SRC_PATTERN[testPatternCount % len(GST_TYPE_VIDEO_TEST_SRC_PATTERN)]
            self.log.info("Test pattern of source '{}' unspecified, picking '{} ({})'"
                          .format(source,pattern, testPatternCount))
        return pattern

    def getTestWave(self, source):
        if not self.has_section('source.{}'.format(source)):
            # background needs no sound, blinder should have no sound
            if source == "blinder" or source == "background":
                return "silence"

        return self.get('source.{}'.format(source), 'wave', fallback="sine")

    def getSourceScan(self, source):
        section = 'source.{}'.format(source)
        if self.has_option(section, 'deinterlace'):
            self.log.error("source attribute 'deinterlace' is obsolete. Use 'scan' instead! Falling back to 'progressive' scheme")
        return self.get(section, 'scan', fallback='progressive')

    def getAudioStreams(self):
        if self.audio_streams is None:
            self.audio_streams = AudioStreams()
            sources = self.getSources()
            for source in sources:
                section = 'source.{}'.format(source)
                if self.has_section(section):
                    self.audio_streams.configure_source(self.items(section), source)
        return self.audio_streams

    def getBlinderAudioStreams(self):
        audio_streams = AudioStreams()
        section = 'source.blinder'
        if self.has_section(section):
            audio_streams += AudioStreams.configure(self.items(section), "blinder", use_source_as_name=True)
        return audio_streams

    def getAudioStream(self, source):
        section = 'source.{}'.format(source)
        if self.has_section(section):
            return AudioStreams.configure(self.items(section), source)
        return AudioStreams()

    def getNumAudioStreams(self):
        num_audio_streams = len(self.getAudioStreams())
        if self.getAudioChannels() < num_audio_streams:
            self.log.error("number of audio channels in mix/audiocaps differs from the available audio input channels within the sources!")
        return num_audio_streams

    def getAudioChannels(self):
        caps = Gst.Caps.from_string(
            self.getAudioCaps()).get_structure(0)
        _, channels = caps.get_int('channels')
        return channels

    def getVideoResolution(self):
        caps = Gst.Caps.from_string(
            self.getVideoCaps()).get_structure(0)
        _, width = caps.get_int('width')
        _, height = caps.get_int('height')
        return (width, height)

    def getVideoRatio(self):
        width, height = self.getVideoResolution()
        return float(width)/float(height)

    def getFramerate(self):
        caps = Gst.Caps.from_string(
            self.getVideoCaps()).get_structure(0)
        (_, numerator, denominator) = caps.get_fraction('framerate')
        return (numerator, denominator)

    def getFramesPerSecond(self):
        num, denom = self.getFramerate()
        return float(num) / float(denom)

    def getVideoSystem(self):
        return self.get('videodisplay', 'system', fallback='gl')

    def getPlayAudio(self):
        return self.getboolean('audio', 'play', fallback=False)

    def getVolumeControl(self):
        # Check if there is a fixed audio source configured.
        # If so, we will remove the volume sliders entirely
        # instead of setting them up.
        return (self.getboolean('audio', 'volumecontrol', fallback=True)
                or self.getboolean('audio', 'forcevolumecontrol', fallback=False))

    def getBlinderEnabled(self):
        return self.getboolean('blinder', 'enabled', fallback=False)

    def isBlinderDefault(self):
        return not self.has_option('blinder', 'videos')

    def getBlinderSources(self):
        if self.getBlinderEnabled():
            if self.isBlinderDefault():
                return ["blinder"]
            else:
                return self.getList('blinder', 'videos')
        else:
            return []

    def getBlinderVolume(self):
        return self.getfloat('source.blinder', 'volume', fallback=1.0)

    def getMirrorsEnabled(self):
        return self.getboolean('mirrors', 'enabled', fallback=False)

    def getMirrorsSources(self):
        if self.getMirrorsEnabled():
            if self.has_option('mirrors', 'sources'):
                return self.getList('mirrors', 'sources')
            else:
                return self.getSources()
        else:
            return []

    def getOutputBuffers(self, channel):
        return self.getint('output-buffers', channel, fallback=500)

    def splitOptions(line):
        if len(line) == 0:
            return None
        quote = False
        options = [""]
        for char in line:
            if char == ',' and not quote:
                options.append("")
            else:
                if char == '"':
                    quote = not quote
                options[-1] += char
        return options

    def getAudioCodec(self, section):
        if self.has_option(section, 'audiocodec'):
            codec = self.get(section, 'audiocodec').split(',',1)
            if len(codec) > 1:
                codec, options = self.get(section, 'audiocodec').split(',',1)
                return codec, VocConfigParser.splitOptions(options) if options else None
            else:
                return codec[0], None
        return "aac", None

    def getAudioEncoder(self, section):
        if self.has_option(section, 'audioencoder'):
            return self.get(section, 'audioencoder')
        return None

    def getVideoCodec(self, section):
        if self.has_option(section, 'videocodec'):
            codec = self.get(section, 'videocodec').split(',',1)
            if len(codec) > 1:
                codec, options = self.get(section, 'videocodec').split(',',1)
                return codec, VocConfigParser.splitOptions(options) if options else None
            else:
                return codec[0], None
        return "jpeg", ["quality=90"]

    def getVideoEncoder(self, section):
        if self.has_option(section, 'videoencoder'):
            return self.get(section, 'videoencoder')
        return None

    def getVideoDecoder(self, section):
        if self.has_option(section, 'decoder'):
            return self.get(section, 'decoder')
        return None

    def getDenoise(self, section):
        if self.has_option(section, 'denoise'):
            if self.getboolean(section, 'denoise'):
                return 1
        return 0

    def getScaleMethod(self, section):
        if self.has_option(section, 'scale-method'):
            return self.getint(section, 'scale-method')
        return 0

    def getDeinterlace(self, section):
        return self.getboolean(section, 'deinterlace', fallback=False)

    def getAudioCaps(self, section='mix'):
        if self.has_option(section, 'audiocaps'):
            return self.get(section, 'audiocaps')
        elif self.has_option('mix', 'audiocaps'):
            return self.get('mix', 'audiocaps')
        else:
            return "audio/x-raw,format=S16LE,channels=2,layout=interleaved,rate=48000"

    def getVideoCaps(self, section='mix'):
        if self.has_option(section, 'videocaps'):
            return self.get(section, 'videocaps')
        elif self.has_option('mix', 'videocaps'):
            return self.get('mix', 'videocaps')
        else:
            return "video/x-raw,format=I420,width=1920,height=1080,framerate=25/1,pixel-aspect-ratio=1/1"

    def getPreviewSize(self):
        width = self.getint('preview', 'width') if self.has_option(
            'preview','width') else 320
        height = self.getint('preview', 'height') if self.has_option(
            'preview','height') else int(width * 9 / 16)
        return(width, height)

    def getLocalPlayoutEnabled(self):
        return self.getboolean('localplayout', 'enabled', fallback=False)

    def getLocalPlayoutAudioEnabled(self):
        return self.getboolean('localplayout', 'audioenabled', fallback=True)

    def getRecordingEnabled(self):
        return self.getboolean('localplayout', 'record', fallback=True)

    def getPreviewsEnabled(self):
        return self.getboolean('previews', 'enabled', fallback=False)

    def getAVRawOutputEnabled(self):
        return self.getboolean('avrawoutput', 'enabled', fallback=True)

    def getLocalUIEnabled(self):
        return self.getboolean('localui', 'enabled', fallback=False)

    def getLocalUIVideoSystem(self):
        return self.get('localui', 'system', fallback="autovideosink")

    def getLivePreviews(self):
        if self.getBlinderEnabled():
            singleval = self.get('previews', 'live').lower()
            if singleval in [ "true", "yes" ]:
                return ["mix"]
            if singleval == "all":
                return self.getLiveSources()
            previews = self.getList('previews', 'live')
            result = []
            for preview in previews:
                if preview not in self.getLiveSources():
                    self.log.error("source '{}' configured in 'preview/live' must be listed in 'mix/livesources'!".format(preview))
                else:
                    result.append(preview)
            return result
        else:
            self.log.warning("configuration attribute 'preview/live' is set but blinder is not in use!")
            return []

    def getComposites(self):
        return Composites.configure(self.items('composites'), self.getVideoResolution())

    def getTargetComposites(self):
        return Composites.targets(self.getComposites())

    def getTransitions(self, composites):
        return Transitions.configure(self.items('transitions'),
                                     composites,
                                     fps=self.getFramesPerSecond())

    def getPreviewNameOverlay(self):
        return self.getboolean('previews', 'nameoverlay', fallback=True)

    def hasSource(self, source):
        return self.has_section('source.{}'.format(source))

    def hasOverlay(self):
        return self.has_section('overlay')

    def getOverlayAutoOff(self):
        return self.getboolean('overlay', 'auto-off', fallback=True)

    def getOverlayUserAutoOff(self):
        return self.getboolean('overlay', 'user-auto-off', fallback=False)

    def getVideoSources(self):
        def source_has_video(source):
            return kind_has_video(self.getSourceKind(source))
        sources = self.getSources()
        return list(filter(source_has_video, sources))

    def getAudioSources(self, internal=False):
        def source_has_audio(source):
            return kind_has_audio(self.getSourceKind(source))
        sources = self.getSources()
        if internal:
            sources += ['mix']
            if self.getBlinderEnabled():
                sources += ['blinder', 'mix-blinded']
        return list(filter(source_has_audio, sources))
